source/geod24/LocalRest.d

```diff
@@ -935,56 +905,33 @@ public final class RemoteAPI (API, alias S = VibeJSONSerializer!()) : API
             mixin(q{
                 override ReturnType!(ovrld) } ~ member ~ q{ (Parameters!ovrld params)
                 {
-                    // we are in the main thread
-                    if (scheduler is null)
-                    {
-                        scheduler = new LocalScheduler;
-                        is_main_thread = true;
-                    }
 
                     // `geod24.concurrency.send/receive[Only]` is not `@safe` but
                     // this overload needs to be
                     auto res = () @trusted {
                         auto serialized = S.serialize(ArgWrapper!(Parameters!ovrld)(params));
+                        auto resp_chn = new RestChannel(4);
+                        auto command = Command(resp_chn, ovrld.mangleof, SerializedData(serialized));
 
-                        auto command = Command(C.thisTid(), scheduler.getNextResponseId(), ovrld.mangleof,
-                                               SerializedData(serialized));
-
-                        // If the node already shut down, its MessageBox will be
-                        // closed. Detect it and notify the user.
-                        // Note that it might be expected that the remote died.
-                        if (!C.trySend(this.childTid, command))
-                            throw new Exception("Connection with peer closed");
+                        Response sendCommandAndWaitResponse ()
+                        {
+                            Variant var;
+                            if(!this.listen_chn.write(Variant(command)))
+                                throw new Exception("Connection with peer closed");
+                            if (resp_chn.read(var, this.timeout))
+                                return var.res;
+                            else
+                                return Response(Status.Timeout, SerializedData("Request timed-out"));
+                        }
```

Comment:
Ah, here it is! That's the main issue: Sending a query to the node should not create a new channel. There should be two kinds of channels:
- 1) When `spawn`ing, you get a `BindChan` (or whatever name you see fit): this is the channel the node listens to for incoming connections.
- 2) When creating a client, you get a `ConnectionChan` (or `Connection`, or whatev) that correspond to the ephemeral connection between a client and a server (what you get after the server has `accept[4]()`ed the connection).

Reply:
Well, my initial motivation was just to get a field test for channels. So I didn't want to re-design localrest around channels. I just replaced the obvious stuff. What you suggest is much better obviously :) I will give it a try.

Reply:
@Geod24 , I gave this approach a shot today and I have some doubts about it. Having a TCP/HTTP like operation using Channels might be more trouble than it's worth. After giving it some thought, the benefits of this is kinda lost on me too :smile:  To me, it seems like only benefit is getting rid of the sender Tid in `Command`. We still have to route responses to whichever Fiber is waiting for it and we now have to manage 4 channels (ie. 2 `Connection`s, since both nodes in a pair can be the server) for **every node pair** in the system instead of 1 `MessageBox` per node.

Reply:
> and we now have to manage 4 channels

4 channels ? Are you using channels as full-duplex connections or do you actually have a read and a write ?

Some aspects must be a bit tricky to manage, for example:
1) The `accept` channel cannot be full duplex, so a node actually need to send a channel or some kind of identifier.
2) Two fibers (client/server) can race on a single connection if used as full-duplex. This is alright as long as there is a way to know who sent/must receive the message, and they are able to put it back in the queue.

> instead of 1 `MessageBox` per node.

The main benefit is to be able to properly type the messages (the previous implementation completely bypasses the type system). There are other smaller benefit (easier to reason about, perhaps ability to control contention), but IMO type-safety is the main one.

Also I still don't see how even 4 channels / node brings overhead as compared to spawning a channel per query. I expect nodes will perform way more than 4 queries during their lifetime. Additionally, the overhead of a node is much higher than what the channel will bring.

Reply:
> 4 channels ? Are you using channels as full-duplex connections or do you actually have a read and a write ?

For async communication, we would need 2 channels. One to send the request, second one to wait the response on. This makes a `Connection`. They get created when a new instance of `RemoteAPI` is created and `Connection` gets sent across the `accept` channel. This `Connection` is only good to send a request to the owner thread of the `RemoteAPI`. Every client should "initiate" a `Connection` to the server by instantiating a `RemoteAPI`, thus 2 `Connections` per every node pair. Using the same `Connection` for passing requests and responses in both directions would cause Nodes to race for initiating the `Connection` and require a mechanism to keep track of already existing `Connections`, but I am not sure how big of a problem that might be. This is almost a complete network stack :smile: 

> 4 channels / node brings

It is actually 4 channels / every possible node pair. 

Overhead of having persistent `Connection`s instead of short living Channels per query is in the implementation complexity. Node have to both manage incoming and outgoing `Connection`s separately. IMO, the biggest drawback of spawning a Channel per query is creating the OS Mutex and the GC allocation of the Channel itself, which I am not sure if either would be a significant bottleneck. But, since we know the lifetime of the response Channel we can allocate it on the stack. Also, OS Mutex can be replaced with a FiberMutex easily.

Essentially what I am trying to say is that I would prefer simplicity over potential performance increase that we probably dont need right now. I might have gotten the whole `Connection` design wrong tho, it might be possible to implement it way simpler than I am trying to right now. :smile:

Reply:
> Using the same `Connection` for passing requests and responses in both directions would cause Nodes to race for initiating the `Connection` and require a mechanism to keep track of already existing `Connections`, but I am not sure how big of a problem that might be.

I'm not sure I follow. So let's say we have a network with 2 nodes, `A` and `B`. 
A` connects as a client to `B`, and `B` connects as a client to `A`.
`A` now have two `Connection`s: the client connection to `B`, and the server `Connection` to `B`. So does `B`.

When nothing happens, nodes listen to their server connection. When they get a query, they process it, and then write back on that same connection.
When doing a request as a client (e.g. during the processing of a query), they first write to the client connection. This should immediately block (yield to the event loop) until the query has been read. Then, they  read from that channel, which blocks (yield to the event loop) until someone writes.

The only moment nodes could ever race is when there's recursion.
E.g. `A [client]` -> `[server] B [client]` -> `[server] A [client]` -> `[server] B`.

However, you can match the  waiting fiber with the response if you associate them with an ID.

Reply:
Moar commits for you to review @Geod24 @AndrejMitrovic  :)